### ISSUE TITLE
CTC loss

### ISSUE THREAD

**Issue by marsiancba** [tensorflow/tfjs#1759]
Please add support for CTC loss.

My use-case is recognizing handwritten numbers (1-5 digits).

**Comment by fahimsun** [comment]
Please add CTC support in Tensorflow JS

CTC is crucial in all cases where Input length is not equal to output length.
For example: Speech Recognition using RNN , handwritten digits in sequence

**Comment by WenheLI** [comment]
Interested in this one, can I have a try with it?

**Comment by rthadur** [comment]
@WenheLI sure please submit a new Pull Request

**Comment by marsiancba** [comment]
@WenheLI I can provide you some rough (but working) implementation of CTC loss in tfjs, if you want something to start from...

**Comment by fahimsun** [comment]
@WenheLI are you working on this?

**Comment by WenheLI** [comment]
@fahimsun Yep, but I am in the break right now, it may take some time to implement. If you want to take it over, it is fine as well.

**Comment by jasonmayes** [comment]
Update: This is blocking new folks work from being published around handwriting recognition and speech recognition. Is there an update for this?

Please see: https://discuss.tensorflow.org/t/ctc-loss-implementation-in-tfjs/6645/5

**Comment by harangp** [comment]
Hi, I've prepared a working CTC loss / gradient calculator for TFJS. Works on batches, handles various length labels for learning, built entirely from scratch based on the original papers. Plugabble into TFJS's model to calculate losses and gradients during model.fit() Open sourced here: https://github.com/harangp/tfjsctcloss Drop me a line if you are interested.

**Comment by gaikwadrahul8** [comment]
Hi, @marsiancba 

Apologize for the delayed response and It seems like at the moment we haven't implemented this feature request so still are you looking this feature ? 

@harangp, Would you like to contribute for this feature ? If yes please refer this [link](https://github.com/tensorflow/tfjs/blob/master/CONTRIBUTING_MISSING_OP.md), Thank you!

**Comment by marsiancba** [comment]
Hi, @gaikwadrahul8 
The part of our project where we used tfjs is stalled atm, but we plan to return to it somewhere in the future.
We used our own implementation (https://github.com/tensorflow/tfjs/issues/1759#issuecomment-534565149) and it seemed to work ok, but native implementation will probably be better/faster.

**Comment by harangp** [comment]
> @harangp, Would you like to contribute for this feature ? If yes please refer this [link](https://github.com/tensorflow/tfjs/blob/master/CONTRIBUTING_MISSING_OP.md), Thank you!

Yes, in time. I still have to work on the backward and collection part to be native, and I'm not confident enough about the returned gradients. I also want to utilize masks to make things more efficient. And lastly, there's a problem with not being a drop-in compatible with the python version, so I'm not sure it would fit into the TFJS concept.

**Comment by nileshau-afk** [comment]
Hi @marsiancba  ,

We are closing this issue due to inactivity. If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.
Thank you for your understanding.
